File: sdk/java/src/test/java/org/finos/openresourcebroker/sdk/parity/ParityScenarioTest.java
package org.finos.openresourcebroker.sdk.parity;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.finos.openresourcebroker.sdk.client.OrbApiException;
import org.finos.openresourcebroker.sdk.client.OrbClient;
import org.finos.openresourcebroker.sdk.model.*;

import org.junit.jupiter.api.*;

import java.io.*;
import java.net.UnixDomainSocketAddress;
import java.nio.channels.SocketChannel;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.*;
import java.util.logging.Logger;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Cross-language parity runner (Java leg).
 *
 * <p>This test LOADS the language-agnostic fixture {@code sdk/parity/scenario.json}
 * and executes its six ordered steps against a REAL orb spawned over a Unix
 * domain socket (the same spawn approach {@code OrbContractTest} uses). Each
 * step is dispatched to the concrete Java SDK method named in the fixture's
 * {@code sdk_methods.java} entry, and the result is asserted against the step's
 * {@code expected} block and skip rules.
 *
 * <p>Static conformance (validate_sdk_spec_conformance.py) proves each step's
 * (method, path, operationId) — and now sdk_methods.java — resolves to a real
 * spec operation and client method. This runtime leg proves the Java SDK
 * actually drives the scenario end-to-end.
 */
@Tag("parity")
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class ParityScenarioTest {

    private static final Logger LOG = Logger.getLogger(ParityScenarioTest.class.getName());
    private static final Duration START_TIMEOUT = Duration.ofSeconds(45);
    private static final String ORB_BINARY = findOrbBinary();
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static Path tmpDir;
    private static String socketPath;
    private static Process orbProcess;
    private static OrbClient client;
    private static JsonNode scenario;

    // Variables bound across steps by the fixture's post_condition rules.
    private static String firstTemplateId;
    private static String requestId;
    private static String machineId;

    private static final Map<Integer, String> RESULTS = new LinkedHashMap<>();

    // ──────────────────────────────────────────────────────────────────────
    // Fixture setup / teardown
    // ──────────────────────────────────────────────────────────────────────

    @BeforeAll
    static void startOrb() throws Exception {
        // Load the shared fixture. build.gradle sets the working dir to sdk/java,
        // so the scenario lives at ../parity/scenario.json.
        Path scenarioPath = Paths.get("..", "parity", "scenario.json");
        assertTrue(Files.exists(scenarioPath), "parity scenario not found: " + scenarioPath.toAbsolutePath());
        scenario = MAPPER.readTree(scenarioPath.toFile());

        tmpDir = Files.createTempDirectory("orb-parity-java-");
        socketPath = tmpDir.resolve("orb.sock").toString();
        Path configPath = tmpDir.resolve("config.json");

        Map<String, Object> config = new LinkedHashMap<>();
        config.put("version", "2.0.0");
        config.put("scheduler", Map.of("type", "default"));
        config.put("provider", Map.of("providers", List.of(Map.of(
                "name", "aws-stub",
                "type", "aws",
                "enabled", true,
                "config", Map.of("region", "us-east-1")))));
        config.put("storage", Map.of("type", "json"));
        config.put("server", Map.of(
                "host", "127.0.0.1",
                "port", 19994,
                "working_dir", tmpDir.toString(),
                "pid_file", tmpDir.resolve("orb.pid").toString()));
        config.put("auth", Map.of("type", "none"));
        config.put("logging", Map.of("level", "ERROR"));
        MAPPER.writeValue(configPath.toFile(), config);

        List<String> cmd = buildOrbCommand(configPath.toString(), socketPath);
        LOG.info("Starting ORB: " + String.join(" ", cmd));
        ProcessBuilder pb = new ProcessBuilder(cmd);
        pb.environment().put("ORB_LOG_LEVEL", "ERROR");
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        pb.redirectError(ProcessBuilder.Redirect.DISCARD);
        orbProcess = pb.start();

        waitForHealthy(socketPath, START_TIMEOUT);

        client = OrbClient.builder()
                .socketPath(socketPath)
                .timeout(Duration.ofSeconds(15))
                .maxRetries(1)
                .retryBaseDelay(Duration.ofMillis(100))
                .build();
    }

    @AfterAll
    static void stopOrb() throws Exception {
        if (scenario != null) {
            for (JsonNode step : scenario.get("steps")) {
                int n = step.get("step").asInt();
                LOG.info("PARITY " + n + " " + step.get("name").asText() + ": " + RESULTS.get(n));
            }
        }
        if (client != null) {
            try { client.close(); } catch (Exception ignored) {}
        }
        if (orbProcess != null) {
            orbProcess.destroy();
            orbProcess.waitFor(5, java.util.concurrent.TimeUnit.SECONDS);
            orbProcess.destroyForcibly();
        }
        if (tmpDir != null) deleteDirectory(tmpDir.toFile());
    }

    private static String methodFor(int step) {
        for (JsonNode s : scenario.get("steps")) {
            if (s.get("step").asInt() == step) {
                return s.get("sdk_methods").get("java").asText();
            }
        }
        return fail("no fixture step " + step);
    }

    // ──────────────────────────────────────────────────────────────────────
    // Steps — one @Test per fixture step, executed in fixture order.
    // ──────────────────────────────────────────────────────────────────────

    @Test
    @Order(1)
    void step1_healthCheck() throws Exception {
        LOG.info("step 1 health_check → " + methodFor(1));
        var result = client.health();
        assertNotNull(result, "health should return a response");
        String status = String.valueOf(result.get("status"));
        assertTrue(status.equals("healthy") || status.equals("degraded"),
                "status should be healthy|degraded, got: " + status);
        RESULTS.put(1, "PASS");
    }

    @Test
    @Order(2)
    void step2_listTemplates() throws Exception {
        LOG.info("step 2 list_templates → " + methodFor(2));
        TemplateListResponse result = client.listTemplates(null, null, null);
        assertNotNull(result.getTemplates(), "templates field must be present (array)");
        if (!result.getTemplates().isEmpty()) {
            firstTemplateId = result.getTemplates().get(0).getTemplateId();
            LOG.info("  bound first_template_id=" + firstTemplateId);
        }
        RESULTS.put(2, "PASS");
    }

    @Test
    @Order(3)
    void step3_requestMachines() throws Exception {
        LOG.info("step 3 request_machines → " + methodFor(3));
        if (firstTemplateId == null) { RESULTS.put(3, "SKIP"); return; }
        var req = new RequestMachinesRequest();
        req.setTemplateId(firstTemplateId);
        req.setCount(1);
        try {
            var result = client.requestMachines(req);
            assertNotNull(result.getRequestId(), "2xx must bind a non-empty request_id");
            requestId = result.getRequestId();
            LOG.info("  bound request_id=" + requestId);
            RESULTS.put(3, "PASS");
        } catch (OrbApiException e) {
            // Provider-level failure (no real AWS) is not a route bug.
            assertNotRouteLevelError(e, "requestMachines");
            LOG.info("  requestMachines non-route error " + e.getStatusCode() + " (expected without real provider)");
            RESULTS.put(3, "SKIP");
        }
    }

    @Test
    @Order(4)
    void step4_pollRequestStatus() throws Exception {
        LOG.info("step 4 poll_request_status → " + methodFor(4));
        if (requestId == null) { RESULTS.put(4, "SKIP"); return; }
        var result = client.getRequestStatus(requestId, null);
        assertNotNull(result, "getRequestStatus should return a response");
        if (result.getRequests() != null && !result.getRequests().isEmpty()) {
            var machines = result.getRequests().get(0).getMachines();
            if (machines != null && !machines.isEmpty()) {
                machineId = machines.get(0).getMachineId();
            }
        }
        RESULTS.put(4, "PASS");
    }

    @Test
    @Order(5)
    void step5_returnMachines() throws Exception {
        LOG.info("step 5 return_machines → " + methodFor(5));
        if (requestId == null || machineId == null) { RESULTS.put(5, "SKIP"); return; }
        var req = new ReturnMachinesRequest();
        req.setMachineIds(List.of(machineId));
        try {
            client.returnMachines(req);
        } catch (OrbApiException e) {
            assertNotRouteLevelError(e, "returnMachines");
            LOG.info("  returnMachines non-route error " + e.getStatusCode() + " (acceptable)");
        }
        RESULTS.put(5, "PASS");
    }

    @Test
    @Order(6)
    void step6_listRequests() throws Exception {
        LOG.info("step 6 list_requests → " + methodFor(6));
        var result = client.listRequests();
        assertNotNull(result, "listRequests should return a response");
        RESULTS.put(6, "PASS");
    }

    // ──────────────────────────────────────────────────────────────────────
    // Helpers (mirror OrbContractTest's harness)
    // ──────────────────────────────────────────────────────────────────────

    private static void assertNotRouteLevelError(OrbApiException e, String context) {
        assertNotEquals(405, e.getStatusCode(),
                context + ": HTTP 405 Method Not Allowed — route-level bug");
        if (e.getStatusCode() == 404) {
            String msg = e.getMessage();
            assertFalse(msg != null && msg.contains("Not Found") && e.getCode() == null
                            && !msg.contains("not found") && !msg.contains("HTTP 404"),
                    context + ": HTTP 404 unstructured 'Not Found' — likely route-level");
        }
    }

    private static String findOrbBinary() {
        String envBinary = System.getenv("ORB_BINARY");
        if (envBinary != null && !envBinary.isEmpty()) return envBinary;
        for (String dir : System.getenv("PATH").split(File.pathSeparator)) {
            File f = new File(dir, "orb");
            if (f.canExecute()) return f.getAbsolutePath();
        }
        for (String py : new String[]{"python3", "python"}) {
            for (String dir : System.getenv("PATH").split(File.pathSeparator)) {
                File f = new File(dir, py);
                if (f.canExecute()) return f.getAbsolutePath();
            }
        }
        return "orb";
    }

    private static List<String> buildOrbCommand(String configPath, String socketPath) {
        List<String> cmd = new ArrayList<>();
        String binary = ORB_BINARY;
        if (binary.endsWith("python3") || binary.endsWith("python")) {
            cmd.add(binary);
            cmd.add("-m");
            cmd.add("orb");
        } else {
            cmd.add(binary);
        }
        cmd.add("--config");
        cmd.add(configPath);
        cmd.add("server");
        cmd.add("start");
        cmd.add("--foreground");
        cmd.add("--api-only");
        cmd.add("--socket-path");
        cmd.add(socketPath);
        return cmd;
    }

    private static void waitForHealthy(String socketPath, Duration timeout)
            throws InterruptedException {
        long deadline = System.currentTimeMillis() + timeout.toMillis();
        Exception lastErr = null;
        while (System.currentTimeMillis() < deadline) {
            Thread.sleep(200);
            try {
                UnixDomainSocketAddress addr = UnixDomainSocketAddress.of(socketPath);
                try (SocketChannel ch = SocketChannel.open(addr)) {
                    ch.configureBlocking(true);
                    var out = java.nio.channels.Channels.newOutputStream(ch);
                    var in = java.nio.channels.Channels.newInputStream(ch);
                    String req = "GET /health HTTP/1.1\r\nHost: localhost\r\nAccept: application/json\r\nConnection: close\r\n\r\n";
                    out.write(req.getBytes(StandardCharsets.US_ASCII));
                    out.flush();
                    BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
                    String statusLine = reader.readLine();
                    if (statusLine != null && statusLine.contains(" 200 ")) {
                        String line;
                        while ((line = reader.readLine()) != null && !line.isEmpty()) {}
                        StringBuilder body = new StringBuilder();
                        char[] buf = new char[512];
                        int n;
                        while ((n = reader.read(buf)) != -1) body.append(buf, 0, n);
                        String b = body.toString();
                        if (b.contains("\"healthy\"") || b.contains("\"degraded\"")) return;
                    }
                }
            } catch (Exception e) {
                lastErr = e;
            }
        }
        throw new IllegalStateException("ORB did not become healthy within " + timeout +
                (lastErr != null ? ": " + lastErr.getMessage() : ""));
    }

    private static void deleteDirectory(File dir) {
        if (dir == null || !dir.exists()) return;
        File[] files = dir.listFiles();
        if (files != null) {
            for (File f : files) {
                if (f.isDirectory()) deleteDirectory(f);
                else f.delete();
            }
        }
        dir.delete();
    }
}

File: sdk/kotlin/src/test/kotlin/org/finos/openresourcebroker/sdk/parity/ParityScenarioTest.kt
/**
 * Cross-language parity runner (Kotlin leg).
 *
 * Loads the language-agnostic fixture sdk/parity/scenario.json and executes its
 * six ordered steps against a REAL ORB spawned over a UNIX domain socket (the
 * same spawn approach ContractTest uses). Each step is dispatched to the
 * concrete Kotlin SDK method named in the fixture's sdk_methods.kotlin entry,
 * and the result is asserted against the step's expected block and skip rules.
 *
 * Static conformance (validate_sdk_spec_conformance.py) proves each step's
 * (method, path, operationId) — and now sdk_methods.kotlin — resolves to a real
 * spec operation and client method. This runtime leg proves the Kotlin SDK
 * actually drives the scenario end-to-end.
 *
 * Required env vars:
 *   ORB_BINARY — absolute path to the orb python binary (e.g. /path/to/.venv/bin/python)
 *   ORB_SRC    — optional path to the orb source root (for PYTHONPATH)
 */

package org.finos.openresourcebroker.sdk.parity

import com.google.gson.Gson
import kotlinx.coroutines.runBlocking
import org.finos.openresourcebroker.sdk.auth.AuthOption
import org.finos.openresourcebroker.sdk.client.ClientConfig
import org.finos.openresourcebroker.sdk.client.OrbApiError
import org.finos.openresourcebroker.sdk.client.OrbClient
import org.finos.openresourcebroker.sdk.model.RequestMachinesRequest
import org.finos.openresourcebroker.sdk.model.ReturnMachinesRequest
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
class ParityScenarioTest {

    private lateinit var client: OrbClient
    private lateinit var proc: Process
    private lateinit var tmpDir: File
    private lateinit var socketPath: String
    private val gson = Gson()

    private lateinit var scenario: Map<*, *>
    private var firstTemplateId: String? = null
    private var requestId: String? = null
    private var machineId: String? = null
    private val results = linkedMapOf<Int, String>()

    companion object {
        private const val START_TIMEOUT_MS = 60_000L
        private val ORB_BINARY: String = System.getenv("ORB_BINARY")
            ?: error(
                "ORB_BINARY environment variable is not set. " +
                "Set it to the absolute path of the orb python binary, e.g.: " +
                "export ORB_BINARY=/path/to/.venv/bin/python"
            )
        private val ORB_SRC = System.getenv("ORB_SRC") ?: ""
    }

    @BeforeAll
    fun startOrb() {
        // build runs from sdk/kotlin, so the fixture is at ../parity/scenario.json.
        val scenarioPath = Paths.get("..", "parity", "scenario.json")
        assertTrue(Files.exists(scenarioPath), "parity scenario not found: ${scenarioPath.toAbsolutePath()}")
        scenario = gson.fromJson(scenarioPath.toFile().readText(), Map::class.java)

        tmpDir = Files.createTempDirectory("orb-parity-kt-").toFile()
        socketPath = "${tmpDir.absolutePath}/orb.sock"
        val configPath = "${tmpDir.absolutePath}/config.json"

        val config = mapOf(
            "version" to "2.0.0",
            "scheduler" to mapOf("type" to "default"),
            "provider" to mapOf(
                "providers" to listOf(
                    mapOf(
                        "name" to "aws-stub",
                        "type" to "aws",
                        "enabled" to true,
                        "config" to mapOf("region" to "us-east-1"),
                    )
                )
            ),
            "storage" to mapOf("type" to "json"),
            "server" to mapOf(
                "host" to "127.0.0.1",
                "port" to 19993,
                "working_dir" to tmpDir.absolutePath,
                "pid_file" to "${tmpDir.absolutePath}/orb.pid",
            ),
            "auth" to mapOf("type" to "none"),
            "logging" to mapOf("level" to "ERROR"),
        )
        File(configPath).writeText(gson.toJson(config))

        val env = ProcessBuilder().environment().toMutableMap()
        if (ORB_SRC.isNotEmpty()) env["PYTHONPATH"] = ORB_SRC
        env["ORB_LOG_LEVEL"] = "ERROR"

        proc = ProcessBuilder(
            ORB_BINARY, "-m", "orb",
            "--config", configPath,
            "server", "start", "--foreground", "--api-only",
            "--socket-path", socketPath,
        )
            .also { it.environment().putAll(env) }
            .redirectErrorStream(false)
            .start()

        Thread { try { proc.inputStream.copyTo(java.io.OutputStream.nullOutputStream()) } catch (_: Exception) {} }
            .also { it.isDaemon = true; it.start() }
        Thread { try { proc.errorStream.copyTo(java.io.OutputStream.nullOutputStream()) } catch (_: Exception) {} }
            .also { it.isDaemon = true; it.start() }

        waitForHealthy()

        client = runBlocking {
            OrbClient.create(
                ClientConfig(
                    socketPath = socketPath,
                    auth = AuthOption.None,
                    timeoutMs = 15_000L,
                )
            )
        }
    }

    @AfterAll
    fun stopOrb() {
        @Suppress("UNCHECKED_CAST")
        val steps = scenario["steps"] as? List<Map<*, *>> ?: emptyList()
        for (step in steps) {
            val n = (step["step"] as Number).toInt()
            println("PARITY $n ${step["name"]}: ${results[n]}")
        }
        if (::client.isInitialized) client.close()
        if (::proc.isInitialized) {
            proc.destroy()
            proc.waitFor(10, java.util.concurrent.TimeUnit.SECONDS)
            if (proc.isAlive) proc.destroyForcibly()
        }
        if (::tmpDir.isInitialized) tmpDir.deleteRecursively()
    }

    private fun methodFor(step: Int): String {
        @Suppress("UNCHECKED_CAST")
        val steps = scenario["steps"] as List<Map<*, *>>
        val s = steps.first { (it["step"] as Number).toInt() == step }
        @Suppress("UNCHECKED_CAST")
        return (s["sdk_methods"] as Map<String, String>)["kotlin"]!!
    }

    // ─────────────────────────────────────────────────────────────────────
    // Steps
    // ─────────────────────────────────────────────────────────────────────

    @Test
    @Order(1)
    fun step1_healthCheck(): Unit = runBlocking {
        println("step 1 health_check -> ${methodFor(1)}")
        val result = client.health()
        assertNotNull(result["status"])
        val status = result["status"] as String
        assertTrue(status in listOf("healthy", "degraded"), "status should be healthy|degraded, got: $status")
        results[1] = "PASS"
    }

    @Test
    @Order(2)
    fun step2_listTemplates(): Unit = runBlocking {
        println("step 2 list_templates -> ${methodFor(2)}")
        val result = client.listTemplates()
        assertNotNull(result.templates, "templates field must be present")
        val templates = result.templates
        if (templates != null && templates.isNotEmpty()) {
            firstTemplateId = templates[0].templateId
            println("  bound first_template_id=$firstTemplateId")
        }
        results[2] = "PASS"
    }

    @Test
    @Order(3)
    fun step3_requestMachines(): Unit = runBlocking {
        println("step 3 request_machines -> ${methodFor(3)}")
        val tid = firstTemplateId
        if (tid == null) { results[3] = "SKIP"; return@runBlocking }
        try {
            val result = client.requestMachines(RequestMachinesRequest(templateId = tid, count = 1))
            assertNotNull(result.requestId, "2xx must bind a non-empty request_id")
            requestId = result.requestId
            println("  bound request_id=$requestId")
            results[3] = "PASS"
        } catch (err: OrbApiError) {
            // Provider-level failure (no real AWS) is not a route bug.
            assertNotRouteLevelError(err, "requestMachines")
            println("  requestMachines non-route error ${err.statusCode} (expected without real provider)")
            results[3] = "SKIP"
        }
    }

    @Test
    @Order(4)
    fun step4_pollRequestStatus(): Unit = runBlocking {
        println("step 4 poll_request_status -> ${methodFor(4)}")
        val rid = requestId
        if (rid == null) { results[4] = "SKIP"; return@runBlocking }
        val result = client.getRequestStatus(rid)
        assertNotNull(result, "getRequestStatus should return a response")
        val requests = result.requests
        if (requests != null && requests.isNotEmpty()) {
            val machines = requests[0].machines
            if (machines != null && machines.isNotEmpty()) {
                machineId = machines[0].machineId
            }
        }
        results[4] = "PASS"
    }

    @Test
    @Order(5)
    fun step5_returnMachines(): Unit = runBlocking {
        println("step 5 return_machines -> ${methodFor(5)}")
        val rid = requestId
        val mid = machineId
        if (rid == null || mid == null) { results[5] = "SKIP"; return@runBlocking }
        try {
            client.returnMachines(ReturnMachinesRequest(machineIds = listOf(mid)))
        } catch (err: OrbApiError) {
            assertNotRouteLevelError(err, "returnMachines")
            println("  returnMachines non-route error ${err.statusCode} (acceptable)")
        }
        results[5] = "PASS"
    }

    @Test
    @Order(6)
    fun step6_listRequests(): Unit = runBlocking {
        println("step 6 list_requests -> ${methodFor(6)}")
        val result = client.listRequests()
        assertNotNull(result, "listRequests should return a response")
        results[6] = "PASS"
    }

    // ─────────────────────────────────────────────────────────────────────
    // Helpers (mirror ContractTest's harness)
    // ─────────────────────────────────────────────────────────────────────

    private fun assertNotRouteLevelError(err: Throwable, context: String) {
        if (err !is OrbApiError) return
        if (err.statusCode == 405) {
            fail<Nothing>("$context: HTTP 405 Method Not Allowed — route-level bug")
        }
        if (err.statusCode == 404 && isRouteLevelNotFound(err)) {
            fail<Nothing>("$context: HTTP 404 with generic 'Not Found' detail — route-level missing path")
        }
    }

    private fun isRouteLevelNotFound(err: OrbApiError): Boolean {
        val detail = parseDetail(err.body)
        return detail != null && detail.trim() == "Not Found"
    }

    private fun parseDetail(body: String?): String? {
        if (body.isNullOrBlank()) return null
        return try {
            val node: Map<*, *>? = gson.fromJson(body, Map::class.java)
            node?.get("detail") as? String
        } catch (_: Exception) {
            null
        }
    }

    private fun waitForHealthy() {
        val deadline = System.currentTimeMillis() + START_TIMEOUT_MS
        while (System.currentTimeMillis() < deadline) {
            Thread.sleep(300)
            if (!proc.isAlive) throw RuntimeException("ORB process exited prematurely")
            try {
                if (pollHealthOverUds()) return
            } catch (_: Exception) {}
        }
        proc.destroyForcibly()
        throw RuntimeException("ORB did not become healthy within ${START_TIMEOUT_MS}ms")
    }

    private fun pollHealthOverUds(): Boolean {
        val factory = org.finos.openresourcebroker.sdk.transport.UdsSocketFactory(socketPath)
        val httpClient = okhttp3.OkHttpClient.Builder()
            .socketFactory(factory)
            .connectTimeout(java.time.Duration.ofMillis(2000))
            .readTimeout(java.time.Duration.ofMillis(2000))
            .build()
        val req = okhttp3.Request.Builder().url("http://localhost/health").get().build()
        return httpClient.newCall(req).execute().use { resp ->
            val body = resp.body?.string() ?: return false
            body.contains("\"status\"")
        }
    }
}
